### ISSUE TITLE
Binding an ObservableCollection to a Collection View

### ISSUE THREAD

**Issue by ramirez** [ReactiveKit/ReactiveKit#20]
Hi,

is there any more extensive example of binding an observableCollection of a collectionView? I'm porting an app I had in SwiftBond to ReactiveKit. I'm trying using the way explained here in the documentation and also using some inspiration from the SwiftBond demo app but I can't seem to be able to figure it out. I either get:

malloc: *** error for object 0x7fa13512a718: incorrect checksum for freed object - object was probably modified after being freed.
*** set a breakpoint in malloc_error_break to debug

from the ReactiveUIKit extension, or

*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Invalid update: invalid number of items in section 0.  The number of items contained in an existing section after the update (29) must be equal to the number of items contained in that section before the update (29), plus or minus the number of items inserted or deleted from that section (29 inserted, 0 deleted) and plus or minus the number of items moved into or out of that section (0 moved in, 0 moved out).'

on the same code. Any ideas or hints?

Thx!

LEo

**Comment by srdanrasic** [comment]
Hi @ramirez, hmm, that sounds bad. Could you post the code related to ObservableCollection and a Collection View and how are using it?

**Comment by ramirez** [comment]
Hi,

my problem seems to be that I keep changing the contents of the collection I'm using, based on an observed value. I have a ViewModel that has:

```var localPosts:ObservableCollection<[Post]> = ObservableCollection<[Post]>([])```

and I observe a variable that triggers a repopulation of the collection (One suspect that I have is that I put that in Init. Maybe I should do that binding later): 

```
init () {
selectedUser.observe { event in
            Account.listPosts()
                .observeNext {
                    value in
                    self.localPosts.removeAll()
                    self.localPosts.insertContentsOf(value, at: 0)
            }
        }
}
```

In my view I bind the observable collection to my collectionView: 
```    
func bindView() {
        viewModel.localPosts.bindTo(collectionView) { (indexPath, dataSource, tableView) -> UICollectionViewCell in
            let cell = self.collectionView.dequeueReusableCellWithReuseIdentifier(reuseIdentifier, forIndexPath: indexPath) as! PostCollectionCellView
            let post = dataSource[indexPath.row]
            let cellViewModel = PostCollectionCellViewModel(post: post)
            cell.viewModel = cellViewModel
            post.updateFromBackend()
            return cell
            }.disposeIn(rBag)
    }
```

That's mostly it. The rest of the code is just layout.

Cheers,

LEo

**Comment by srdanrasic** [comment]
Hmm, we might have a bug in the binding...

What happens if you replace 

```swift
self.localPosts.removeAll()
self.localPosts.insertContentsOf(value, at: 0)
```

with

```swift
self.localPosts.replace(value)
```

That basically does the same thing, just in one line.

I'll try to reproduce modifying the collection before doing the binding - but that should also work.

**Comment by ramirez** [comment]
I tried but I get the same. The only way I found to make it work is by doing the binding before ever touching the collection.

**Comment by srdanrasic** [comment]
Hi @ramirez, I've traced the issue to table/collection view binding. Just released ReactiveUIKit 1.0.9 with the issue fixed. Please confirm if it works. Thanks!

**Comment by ramirez** [comment]
Works. Thx!

LEo

**Comment by retsohuang** [comment]
I use ReactiveUIKit 1.0.9 unfortunately it crashes after I update the data
Here's log
```
*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'attempt to insert item 8 into section 0, but there are only 8 items in section 0 after the update'
```

And I use the same function `self.localPosts.replace(value)`

**Comment by srdanrasic** [comment]
Looks like I have to spend some more time on this :) @retsohuang is it collection or table view?

**Comment by retsohuang** [comment]
That issue appears both CollectionView and TableView.

**Comment by srdanrasic** [comment]
Hi @retsohuang, could you provide some more context around the issues? I'm having hard time reproducing this...

**Comment by retsohuang** [comment]
I'll try to create demo project that reproduce this issue.

**Comment by AnthonyMDev** [comment]
This looks like its related to ReactiveKit/ReactiveUIKit#3.

Try removing the `skip(1)` as is described in that issues and see if that fixes the issue like it did for me.

**Comment by AnthonyMDev** [comment]
I'm 99% sure that this is fixed by ReactiveKit/ReactiveUIKit#5. I'd love to get this pulled in ASAP @srdanrasic 
Thanks!
